nemo/collections/tts/modules/submodules.py

```diff
@@ -401,3 +413,280 @@ def forward(self, forward_input: Tuple[torch.Tensor, torch.Tensor]):
+        if token_id is not None:
+            scores = torch.zeros_like(scores)
+            scores[:, :, token_id] = 1
+
+        style_emb = torch.bmm(scores, v).squeeze(1)
+        style_emb = style_emb.contiguous().view(self.n_head, bs, self.token_size)
+        style_emb = style_emb.permute(1, 0, 2).contiguous().view(bs, -1)
+
+        return style_emb
+
+
+class SVEmbeddingLinearProjectionNetwork(torch.nn.Module):
+    def __init__(self, speaker_embedding_dim, symbols_embedding_dim):
+        super(SVEmbeddingLinearProjectionNetwork, self).__init__()
+        self.speaker_proj = torch.nn.Linear(speaker_embedding_dim, symbols_embedding_dim)
+        self.bn1 = torch.nn.BatchNorm1d(num_features=symbols_embedding_dim)
+        self.act_fn = torch.nn.ReLU()
+
+    def forward(self, sv_embedding):
+        x = self.speaker_proj(sv_embedding)
+        return self.act_fn(self.bn1(x)).unsqueeze(1)
+
+
+
+class SpeakerEncoder(torch.nn.Module):
+    """
+    class SpeakerEncoder represents speakers representation. This module can combine GST (global style token)
+    based speaker embeddings, speaker embeddings from speaker verificaiton models and lookup table speaker
+    embeddings.
+    """
+    def __init__(self, gst_module=None, sv_projection_module=None, lookup_emb_projection_module=None):
+        """
+        Constructor.
+        Parameters:
+            - gst_module: Neural module to get GST based speaker embedding
+            - sv_projection_module: Neural module to get Speaker Verification based speaker embedding
+            - lookup_emb_projection_module: Neural module to project lookup speaker embedding
+        """ 
+        super(SpeakerEncoder, self).__init__()
+        self.gst_module = gst_module
+        self.sv_projection_module = sv_projection_module
+        self.lookup_emb_projection_module = lookup_emb_projection_module
+
+    def forward(self, spk_emb, gst_ref_spec=None, gst_ref_spec_lens=None, speaker_embedding=None):
+        x = spk_emb
+        # Project lookup speaker embedding
+        if self.lookup_emb_projection_module is not None:
+            x += self.lookup_emb_projection_module(spk_emb)
+
+        # Get GST based speaker embedding
+        if self.gst_module is not None and gst_ref_spec is not None and gst_ref_spec_lens is not None:
+            ref_spec_mask = (
+                torch.arange(gst_ref_spec_lens.max())
+                .to(gst_ref_spec.device)
+                .expand(gst_ref_spec_lens.shape[0], gst_ref_spec_lens.max())
+                < gst_ref_spec_lens.unsqueeze(1)
+             ).unsqueeze(2)
+            x += self.gst_module(gst_ref_spec, ref_spec_mask).unsqueeze(1)
+
+        # Project speaker embedding from Speaker Verification based 
+        if self.sv_projection_module is not None and speaker_embedding is not None:
+            x += self.sv_projection_module(speaker_embedding)
+        return x
+
+
+class Conv2d(torch.nn.Module):
+    """
+    Convolution 2D Module
+    """
```

Comment:
Could use a more unique name and clarification/comment that this wrapper is intended to handle the boilerplate transposes needed to convert input dimension X to conv2d input dimensions Y.